lib/pos_mempool.go

```diff
@@ -0,0 +1,446 @@
+	// buckets is the data structure storing the transactions in fee-time order.
+	buckets *FeeBucket
+	// txnMembership is a map of transaction hashes. It is used to determine existence of a transaction in the register.
+	txnMembership map[BlockHash]struct{}
+	// totalTxnSize is the total size of all transactions in the register.
+	totalTxnSize uint64
+
+	params       *DeSoParams
+	globalParams *GlobalParamsEntry
+}
+
+func NewTransactionRegister(params *DeSoParams, globalParams *GlobalParamsEntry) *TransactionRegister {
+	buckets := NewFeeBucket(globalParams)
+	return &TransactionRegister{
+		buckets:       buckets,
+		txnMembership: map[BlockHash]struct{}{},
+		totalTxnSize:  0,
+		params:        params,
+		globalParams:  globalParams,
+	}
+}
+
+// AddTransaction adds a transaction to the register. If the transaction already exists in the register, or its size
+// exceeds the maximum mempool capacity, it is not added.
+func (tr *TransactionRegister) AddTransaction(txn *MempoolTx) bool {
+	if txn.Hash == nil {
+		return false
+	}
+
+	if _, ok := tr.txnMembership[*txn.Hash]; ok {
+		return false
+	}
+
+	// If the transaction overflows the maximum mempool size, reject it.
+	if tr.totalTxnSize+txn.TxSizeBytes > tr.params.MaxMempoolPosSizeBytes {
+		return false
+	}
+
+	if err := tr.buckets.AddTransaction(txn); err != nil {
+		glog.Errorf("TransactionRegister.AddTransaction: Problem adding txn: %v", err)
+		return false
+	}
+	tr.totalTxnSize += txn.TxSizeBytes
+	tr.txnMembership[*txn.Hash] = struct{}{}
+	return true
+}
+
+// RemoveTransaction removes a transaction from the register.
+func (tr *TransactionRegister) RemoveTransaction(txn *MempoolTx) bool {
+	if _, ok := tr.txnMembership[*txn.Hash]; !ok {
+		return false
+	}
+
+	tr.buckets.RemoveTransaction(txn)
+	delete(tr.txnMembership, *txn.Hash)
+	tr.totalTxnSize -= txn.TxSizeBytes
+	return true
+}
+
+func (tr *TransactionRegister) Empty() bool {
+	return tr.buckets.Empty()
+}
+
+// GetFeeTimeIterator returns an iterator over the transactions in the register. The iterator goes through all transactions
+// as ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeIterator() *FeeTimeIterator {
+	return tr.buckets.GetFeeTimeIterator()
+}
+
+// GetFeeTimeTransactions returns all transactions in the register ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeTransactions() []*MempoolTx {
+	txns := []*MempoolTx{}
+	it := tr.GetFeeTimeIterator()
+	for it.Next() {
+		if txn, ok := it.Value(); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+// ========================
+//	FeeBucket
+// ========================
+
+// FeeBucket is a data structure storing time buckets. Time buckets, in turn, store timestamp-ordered mempool transactions
+// that have neighbouring fees. When a transaction is added to the FeeBucket, an appropriate TimeBucket is found or created,
+// depending on the transaction's fee, and the transactions is inserted into it.
+type FeeBucket struct {
```

Comment:
Do we really need a separate FeeBucket data structure? It seems like you can simplify things by merging the FeeBucket data structure with the TransactionRegister data structure. Namely, my suggestion would be that what we actually want is two data structures only:
- A FeeTimeBucket data structure that's identical to your existing TimeBucket struct. This construct  knows its min/max fee rates, and maintains a time ordered treeset of txns for a single fee time ordered bucket
- A TransactionRegister data structure, which maintains a treeset of FeeTimeBuckets ordered by fee bucket

It makes sense that why you would wanted to break things up into three separate structs. But while reading the code, the middle FeeBucket struct ended up actually obfuscating what was happening. Collapsing things into 2 data structures, and updating TimeBucket name to FeeTimeBucket would make it more intuitive to the reader how txn storage is organized.

Reply:
Yup! I think I replied to this off-gh, so yeah, I totally agree.